KEP-3015: PreferLocal traffic policy

danwinship opened the pull request:
- One-line PR description:
Addition of `PreferLocal` (ie, `Local` with fallback) to `externalTrafficPolicy` and `internalTrafficPolicy`

<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/3015

<!-- other comments or additional information -->
- Other comments:
based on discussion in #3010, which I'll close now

danwinship commented:
Updated:

  - don't say "guarantee"
  - totally rewrote the bullet points at the end of the "External Traffic Policy" section to better describe the benefits and tradeoffs (also without saying "guarantee")
  - added a section on how it interacts with topology

robscott commented on keps/sig-network/3015-prefer-local-traffic-policy/README.md:
```diff
@@ -0,0 +1,535 @@
+(FWIW, the expected situation when `PreferLocal` has to fall back to
+sending to a remote endpoint is when pods are in flux, and some nodes
+that are "supposed to" have an endpoint don't yet/anymore. But when
+that happens, there's a good chance that the topology code will _also_
+hit its fallback case (due to there being at least one zone which is
+missing an endpoint), and so topology won't even get used anyway. At
```
I'm not sure this is true. For example, there may be a Service with 3 endpoints spread over 6 nodes. That would mean "Local" would fallback to "Cluster" ~1/2 of the time, but topology would consistently be helpful. Having a "PreferLocal" mode may result in a larger set of users than the traditional "Local" use cases, including some cases where users are not intending to deploy workloads on every possible node.

danwinship replied:
> there may be a Service with 3 endpoints spread over 6 nodes. That would mean "Local" would fallback to "Cluster" ~1/2 of the time

(I assume you meant `PreferLocal` would fall back to `Cluster`.)

> including some cases where users are not intending to deploy workloads on every possible node.

Hm... I was not considering that as a use case, although it does make sense. It's the same behavior, you're just thinking of it as "`Cluster` with optimization" rather than "`Local` with fallback". I should add a User Story for that.

At any rate, the comment here about `PreferLocal` and Topology failing at the same time was purely "informational" and doesn't affect the actual behavior; the proposed rule ("use a local endpoint if you can, otherwise do exactly what you would have done if the policy was `Cluster`") will give the right answer whether or not Topology is in effect.

astoycos commented:
/cc

danwinship commented:
Updated:
- added new User Story "`PreferLocal` as More-Efficient `Cluster`"
- removed incorrect parenthetical comment about PreferLocal vs Topology
- updated the "SLO" table at the end for the new use case

andrewsykim commented on keps/sig-network/3015-prefer-local-traffic-policy/README.md:
```diff
@@ -0,0 +1,548 @@
+pod, for efficiency. However, if no local DNS pod is available, DNS
+should just go to a remote pod instead so it keeps working.
+
+(This is a modified version of a user story from the [Internal Traffic
```
The original KEP actually proposed PreferLocal as one of the policy types, but we later removed it because we thought topology-aware routing could provide the same behavior with less complexity. Curious what changed for us to recondier PreferLocal again?  

(sorry if we already discussed this somewhere else, catching up on some context I missed)

andrewsykim replied:
Sorry, spoke too soon! I see the section "### Interaction with Topology" now, will leave commens there

andrewsykim commented on keps/sig-network/3015-prefer-local-traffic-policy/README.md:
```diff
@@ -0,0 +1,548 @@
+ignore topology, just like `Local` policy does.
+
+When delivering packets remotely, it makes the most sense to have
+`PreferLocal` policy make use of topology. (This is explicitly wanted
```
One of the reasons we scrapped PreferLocal in the other KEP is because the assumption was you can achieve the same behavior by setting internalTrafficPolicy: Cluster and topology-aware routing that strongly prefers node-local endpoints. What is the semantic difference between that and using PreferLocal?

danwinship replied:
If we had node-level topology, then we probably would not need `PreferLocal`... but do we even have a plan for node-level topology yet?

danwinship replied:
Arguably if we had node-level topology, then we would not need `internalTrafficPolicy: Local` either, since (unlike `externalTrafficPolicy: Local`) it provides no advantages other than locality.

thockin replied:
I'll attach my main feedback here.

Let's consider: Why do people use [EI]TP ?

ETP:  I know of 2 reasons.  1) "I _need_ to preserve source IP".  2) "Efficiency".

ITP: I know of 2 reasons.  1) "I _need_ to stay on this node (semantic correctness)".  2) "Efficiency".

For both classes of traffic, "Local" is either a hard requirement (i.e. "prefer" is a non-starter) or just an optimization over "Cluster".

I don't like that we put the responsibility onto our users tell us which optimizations we should be using.  That seems like a total failure on our part.  OF COURSE they want us to optimize if we can.  The specific optimization that PreferLocal dictates is pretty black or white, which means there's not much room for being "smart".  This then FORCES other implementations, which might be capable of more smartness, to play dumb.

For example, it's really not hard to imagine an implementation that sends the number of active flows per-endpoint to an aggregator, and then have that aggregator give the node-agents hints about which endpoints are overloaded.  It doesn't even need to be super timely to have some impact, I bet.  But the semantics of "PreferLocal" disallow that.  

If we think this is the optimizatrion with the biggest bang for the buck, let's talk about how we can do something like `if num_endpoints > (num_nodes/2)  && have_local_eps() { use_local_eps() }`.  I just made that up - obviously we can tune the heuristic.

terinjokes replied:
For our use cases, mostly daemonsets providing platform infrastructure, we definitely have more _prefer_ local endpoints than _needs_ local endpoints. The current semantics for ITP "Local" is very bad for our users, as the daemonset pod becoming not ready provides not optimization but a production outage (in a form of dns cache, for example). Conversely, ITP "Cluster" and Topology Aware Hints also provide the wrong semantics, where we're more likely to not send traffic locally.

We've had to leave ITP as "Cluster" and have external tools try to update iptables when the daemonset pods are ready, and remove when they are not, which is a fragile situation we'd prefer not to be in. We were looking forward to PreferLocal in the previous proposal, as not having it causes us pain today.

Many of our users, who are running business workflows in deployments, prefer semantics of "Cluster" and Topology Aware Hints. Trying to be "smart" can potentially hotspot specific instances just because they are on the same node as an ingress controller or another workload.

> If we think this is the optimizatrion with the biggest bang for the buck, let's talk about how we can do something like `if num_endpoints > (num_nodes/2) && have_local_eps() { use_local_eps() }`.

I worry a heuristic along this pattern would be surprising to our users. Currently, most aspects of scaling clusters up and down are transparent operations to them. These types of heuristics would suddenly change traffic patterns for them, in potentially unexpected ways.

aojea replied:
@thockin just to see if I get your point, you mean that PreferLocal is a loadbalancing/topology algorithm problem and not a traffic policy problem, hence this is not the right place to fix it?

thockin replied:
> PreferLocal is a loadbalancing/topology algorithm problem and not a traffic policy problem, hence this is not the right place to fix it

@aojea That's a reasonable distillation :)

@terinjokes thanks for the additional input.  "Local" gets better with the terminating endpoints work that is in flight, but still there's a moment when it has to die for a successor to start.  If we really want to make node-level "sidecars" a thing that is reliable, we need a way to "hand off".

But such a mechanism does (kind of) exist - the `updateStrategy` of a DaemonSet allows `maxSurge` to be positive, so we can start a new replica before killing an old one.  This is subject to resource limitations on heavily packed nodes.  If you think that is you, you can choose to run 2 DaemonSets (e.g. "dns-1" and "dns-a") so you pre-allocate space.  Maybe DaemonSet actually needs "replicas" field to run N copies per target node...

Topology, as implemented today, is not really considering node.  It could, and yes my proposed heuristic is far from  perfect.  But not considering node was sort of intentional.  In my experience, such optimizations quietly become functional requirements.  Apps (or SLOs) start to _depend_ on the optimization, and when the optimization isn't available for some reason, apps break.  

So my intuitive lean is to either NOT optimize that way, or else to make it really clear that it is an optimization, not a policy.  If you _depend_ on the optimization, then "prefer" is wrong, and we should focus on making "Local" work for you.  If you don't depend on it, then you should let the system try to be smart, and not tell it how to do its job.

That said I really do understand how "PreferLocal" is attractive, especially while the system is NOT as smart as you want it to be.  On this topic, I have cited this before: https://en.wikipedia.org/wiki/Attractive_nuisance_doctrine .  As you said `Trying to be "smart" can potentially hotspot specific instances`.  It's an overly-sharp tool that SEEMS like a good idea, but probably isn't. 

In some sense what you are asking for is not an abstract policy, but an implementation-centric toggle that says "I know what I am doing and I know how the implementation works".  If we put that into the API we get to support it FOREVER.  We reduce the degrees of freedom implementations have for doing good things, and we lock them into sub-optimal semantics.

danwinship replied:
> ITP: I know of 2 reasons. 1) "I need to stay on this node (semantic correctness)". 2) "Efficiency".

`internalTrafficPolicy: Local` doesn't solve the "efficiency" use case though, because it implies "I would rather lose packets than let this connection go to another node", which is never what you want for the "efficiency" use case.

> I don't like that we put the responsibility onto our users tell us which optimizations we should be using. That seems like a total failure on our part. OF COURSE they want us to optimize if we can.

But optimize for _what_?

"Cluster" traffic policy optimizes under the assumption that your service will work best if kube-proxy distributes connections evenly across all the endpoints. This might be the best choice if each connection creates a lot of load on the endpoint, so overloading a local endpoint would slow response time down by more than forwarding requests to another node would.

"PreferLocal" traffic policy optimizes under the assumption that the service will work best if kube-proxy routes connections to the closest endpoint. This is the right choice for something like DNS where connections are short lived, and processing them doesn't require much CPU, so it doesn't matter if lots of connections get sent to the same endpoint all at once.

> The specific optimization that PreferLocal dictates is pretty black or white, which means there's not much room for being "smart".

Yeah, because actually I lied above: "PreferLocal" doesn't say "route connections to the closest endpoint", it says "route connections to a local endpoint, or if you can't, then route it anywhere", which is better than "Cluster"'s behavior, but not really exactly what you want.

> This then FORCES other implementations, which might be capable of more smartness, to play dumb.

In theory an implementation might be super smart and actively monitor the cluster to figure out the best way to route connections, but I think it's more likely that you'll get good behavior if we let the user give the proxy implementation hints. We just need to figure out the right set of hints.

I guess having traffic policy encode both behavioral changes ("preserve source IP", "must stay on node for semantic correctness") _and_ efficiency hints is probably a bad approach. But in that case we should clarify that traffic policy is only describing desired behavior, not desired efficiency.

thockin replied:
I know of cases where falling back to a non-local endpoint is semantically wrong, and isn't acceptable.  Setting a policy to "Local" ought to be a policy, not a hint.  *IF* we support "PreferLocal", it has to be disticnt from "Local".  It still seems wrong to conflate policy and efficiency hints.  

E.g. imagine a near-ideal implementation.  It would gather stats about how many connections and bytes were sent per usint time.  Maybe it would consider latency (cracking flows to recognize DNS or maybe just based on ports).  Maybe it would consider CPU load on the server.  It would use that to make a "good" choice.  Hopefully "pods on this node" would be the choice whenever possible.  We don't need a user to tell us that.

Now, we don't have a near-ideal implementation, so should we:

a) give a policy knob that a closer-to-ideal solution might wrestle with

b) give a policy knob that other implementations can choose to ignore (e.g. annotation);

c) let the user offer a hint that is explicitly a hint (which implementations can choose to respect or not)

d) start building a more ideal implementation

e) ignore this problem and hope it goes away

aojea replied:
> I know of cases where falling back to a non-local endpoint is semantically wrong, and isn't acceptable. Setting a policy to "Local" ought to be a rule

Fair enough

b) seems the simplest solution

aojea commented on keps/sig-network/3015-prefer-local-traffic-policy/README.md:
````diff
@@ -0,0 +1,548 @@
+```
+<<[UNRESOLVED renaming ]>>
+
+Should we add a new alias for `Local`, such as `PreserveSourceIP`, and
````
:) https://github.com/kubernetes/kubernetes/pull/41162#discussion_r102340291

aojea commented on keps/sig-network/3015-prefer-local-traffic-policy/README.md:
```diff
@@ -0,0 +1,548 @@
+
+The KEP gives the use case:
+
+    As a platform owner, I want to create a Service that always
+    directs traffic to a logging daemon on the same node. Traffic
+    should never bounce to a daemon on another node.
```
Singleton pattern? I honestly tried to find an example and I couldn't

danwinship replied:
I'm not sure what you're saying.

aojea replied:
this can be useful if you want to have a service with special characteristics per node... I can't find an example neither, I was thinking if maybe someone want to have an unique service in the cluster that only works in one node ...

thockin reviewed (commented):
Overall I am -1 on this proposal.  Rationale in a line-comment below.

danwinship commented:
I'm going to close this PR and push a new more topology-focused version
